### PR TITLE
Data loader (sampler component) - Kafka/Kinesis samplers

### PR DESCRIPTION
**Pull request by dclim**
Implementation of the sampler component of #7502.
Depends on #7531. 

Adds additional implementations to support sampling from Kafka and Kinesis.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      }
+
+      closed = true;
+      recordSupplier.close();
+    }
+
+    private void assignAndSeek() throws InterruptedException
+    {
+      final Set<StreamPartition<PartitionIdType>> partitions = recordSupplier
+          .getPartitionIds(ioConfig.getStream()).stream()
```
Comment (clintropolis) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java]: nit: .stream() should probably be on newline

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+  protected abstract Firehose getFirehose(InputRowParser parser);
+
+  protected abstract class SeekableStreamSamplerFirehose implements Firehose
+  {
+    private final InputRowParser parser;
+    private final RecordSupplier<PartitionIdType, SequenceOffsetType> recordSupplier;
+
+    private Iterator<OrderedPartitionableRecord<PartitionIdType, SequenceOffsetType>> recordIterator;
+    private Iterator<byte[]> interRecordIterator;
```
Comment (clintropolis) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java]: the starting bit of `nextRowWithRaw` might be a bit clearer if this variable was named something like `recordBytesIterator` or `recordDataIterator`?

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    this.ioConfig = Preconditions.checkNotNull(ingestionSpec.getIoConfig(), "[spec.ioConfig] is required");
+    this.tuningConfig = ingestionSpec.getTuningConfig();
+    this.samplerConfig = samplerConfig;
+    this.firehoseSampler = firehoseSampler;
+  }
+
+  @Override
+  public SamplerResponse sample()
+  {
+    return firehoseSampler.sample(new FirehoseFactory()
```
Comment (clintropolis) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamSamplerSpec.java]: nit: formatting

**Comment by vogievetsky** [comment]
Been testing and using this feature as a "user", seems to work really well. Check it out here: https://youtu.be/tAEp5BXVHYE

**Pull request merged**